examples/DataApi-Voting-poll/template/modules/shared_data/src/main/scala/com/my/currency/shared_data/Main.scala

```diff
@@ -0,0 +1,351 @@
+    signedUpdate.value match {
+      case newPoll: NewPoll =>
+        val validateAddress = signedUpdate.proofs
+          .map(_.id)
+          .toList
+          .traverse(_.toAddress[IO])
+          .map(validateProvidedAddress(_, newPoll.owner))
+
+        val validatePoll = IO {
+          val voteId = Hash.fromBytes(customUpdateSerialization(newPoll))
+          validateIfNewPollExists(oldState.polls.get(voteId.toString))
+        }
+
+        for {
+          validatedAddress <- validateAddress
+          validatedPoll <- validatePoll
+        } yield validatedAddress.productR(validatedPoll)
+
+      case pollVote: PollVote =>
+        val validateAddress = signedUpdate.proofs
+          .map(_.id)
+          .toList
+          .traverse(_.toAddress[IO])
+          .map(validateProvidedAddress(_, pollVote.address))
+
+
+        val validatePoll = IO {
+          validateIfVotePollExists(oldState.polls.get(pollVote.id))
+        }
+
+        val validateOption = IO {
+          validateIfOptionExists(oldState.polls.get(pollVote.id), pollVote.option)
+        }
+
+        val validateRepeatedVote = IO {
+          validateIfUserAlreadyVoted(oldState.polls.get(pollVote.id), pollVote.address)
+        }
+
+        for {
+          validatedAddress <- validateAddress
+          validatedPoll <- validatePoll
+          validatedOption <- validateOption
+          validatedRepeatedVote <- validateRepeatedVote
+        } yield validatedAddress.productR(validatedPoll).productR(validatedOption).productR(validatedRepeatedVote)
+    }
+  }
+}
+
+object Utils {
+  def customUpdateSerialization(update: PollUpdate): Array[Byte] = {
+    println("Serialize UPDATE event received")
+    println(update.asJson.deepDropNullValues.noSpaces)
+    update.asJson.deepDropNullValues.noSpaces.getBytes(StandardCharsets.UTF_8)
+  }
+
+  def customStateSerialization(state: State): Array[Byte] = {
+    println("Serialize STATE event received")
+    println(state.asJson.deepDropNullValues.noSpaces)
+    state.asJson.deepDropNullValues.noSpaces.getBytes(StandardCharsets.UTF_8)
+  }
+
+  def customStateDeserialization(bytes: Array[Byte]): Either[Throwable, State] = {
+    parser.parse(new String(bytes, StandardCharsets.UTF_8)).flatMap { json =>
+      json.as[State]
+    }
+  }
+
+  def customUpdateDeserialization(bytes: Array[Byte]): Either[Throwable, PollUpdate] = {
+    parser.parse(new String(bytes, StandardCharsets.UTF_8)).flatMap { json =>
+      json.as[PollUpdate]
+    }
+  }
+}
+
+object Data {
+  @derive(decoder, encoder)
+  sealed trait PollUpdate extends DataUpdate {
+    val timestamp: Long
+  }
+
+  @derive(decoder, encoder)
+  case class NewPoll(name: String, owner: Address, pollOptions: List[String], startSnapshotOrdinal: Long, endSnapshotOrdinal: Long, timestamp: Long) extends PollUpdate
+
+  @derive(decoder, encoder)
+  case class PollVote(id: String, address: Address, option: String, timestamp: Long) extends PollUpdate
```

Comment:
Maybe `voteInPollUpdate` is better?

Also, is `id` here referring to the ID of the poll? or of the vote? If it's of the poll then I think it should be `pollId` instead of `id`.

Reply:
Changed to pollId